Add Migration Guide & Update DropInRequest for Migration Guide Changes (#260)
braintree/braintree-android-drop-in: v6_MIGRATION_GUIDE.md

# Braintree Android Drop-In v6 Migration Guide

See the [CHANGELOG](/CHANGELOG.md) for a complete list of changes. This migration guide outlines the basics for updating your Braintree Drop-In integration from v5 to v6.

> v6 of the Drop-In SDK is compatible with v4 of the [Braintree Android SDK](https://github.com/braintree/braintree_android).

Reviewer: ```suggestion
v6 of the Drop-In SDK requires v4 of the [Braintree Android SDK](https://github.com/braintree/braintree_android).
```

take it or leave it suggestion

_Documentation for v6 will be published to https://developer.paypal.com/braintree/docs once it is available for general release._

## Table of Contents

1. [Gradle](#gradle)
1. [Builder Pattern](#builder-pattern)
1. [DropInRequest](#dropinrequest)
1. [Authorization](#authorization)
1. [Launch Drop-In](#launch-drop-in)
1. [Handle Drop-In Result](#handle-drop-in-result)
1. [Fetch Last Used Payment Method](#fetch-last-used-payment-method)

## Gradle

Add the dependency in your `build.gradle`:

```groovy
dependencies {
  implementation 'com.braintreepayments.api:drop-in:6.0.0'
}
```

The credentials for accessing the Cardinal Mobile SDK have changed.

Update the credentials in your top-level build.gradle to the following:

```groovy
repositories {
    maven {
        url "https://cardinalcommerceprod.jfrog.io/artifactory/android"
        credentials {
            username 'braintree_team_sdk'
            password 'AKCp8jQcoDy2hxSWhDAUQKXLDPDx6NYRkqrgFLRc3qDrayg6rrCbJpsKKyMwaykVL8FWusJpp'
        }
    }
}

> ```

Reviewer: Irrelevant of the major version bump - should these new credentials also be in our README for braintree_android?

Reviewer 2: That might make sense yeah. They are in the [MIGRATION_GUIDE](https://github.com/braintree/braintree_android/blob/master/v4_MIGRATION_GUIDE.md#3d-secure).

Author: Agreed - will change that outside of this PR

## Builder Pattern

The builder pattern has been removed in v6 to allow for consistent object creation across Java and Kotlin. 
Method chaining has been removed, and setters have been renamed with the `set` prefix.

> For example, a `DropInRequest` can now be constructed below:

Reviewer: ```suggestion
For example, a `DropInRequest` can now be constructed as shown below:
```

Java:
```java
DropInRequest request = new DropInRequest();
request.setCollectDeviceData(true);
request.setMaskCardNumber(true);
```

> Kotlin:

Author: I added both Kotlin and Java snippets here to make it easier to write developer docs for this version, but they could be removed since we don't have Kotlin snippets in the Android core v4 migration guide.

Reviewer: Great idea! I like having both.

## DropInRequest

The getters and setters on `DropInRequest` have been renamed with a consistent get/set pattern to allow for Kotlin synthesized properties. 
See the examples below for a full list of optional parameters:

Java:
```java
    DropInRequest dropInRequest = new DropInRequest();
    dropInRequest.setShouldCollectDeviceData(true);
    dropInRequest.setGooglePayRequest(googlePayRequest);
    dropInRequest.setGooglePayDisabled(true);
    dropInRequest.setPayPalRequest(paypalRequest);
    dropInRequest.setPayPalDisabled(true);
    dropInRequest.setVenmoDisabled(true);
    dropInRequest.setCardDisabled(true);
    dropInRequest.setShouldRequestThreeDSecureVerification(true);
    dropInRequest.setThreeDSecureRequest(threeDSecureRequest);
    dropInRequest.setShouldMaskCardNumber(true);
    dropInRequest.setShouldMaskSecurityCode(true);
    dropInRequest.setEnableVaultManager(true);
    dropInRequest.setAllowVaultCardOverride(true);
    dropInRequest.setVaultCardDefaultValue(true);
    dropInRequest.setCardholderNameStatus(CardForm.FIELD_OPTIONAL);
    dropInRequest.setVaultVenmoDefaultValue(true);
```

Kotlin:
```kotlin
    val dropInRequest = DropInRequest()

>     dropInRequest.shouldCollectDeviceData = true

Reviewer: In iOS Drop In - we are always returning device data now on the DropInResult. So we don't have this param on the DropInRequest. We can also make this change for Android Drop In v6 while we have the luxury of making breaking changes 👯

Author: Nice ok I'll create a JIRA card for that change before GA 👍

    dropInRequest.googlePayRequest = googlePayRequest
    dropInRequest.googlePayDisabled = true
    dropInRequest.payPalRequest = paypalRequest
    dropInRequest.payPalDisabled = true
    dropInRequest.venmoDisabled = true
    dropInRequest.cardDisabled = true
    dropInRequest.shouldRequestThreeDSecureVerification = true
    dropInRequest.threeDSecureRequest = threeDSecureRequest
    dropInRequest.shouldMaskCardNumber = true
    dropInRequest.shouldMaskSecurityCode = true
    dropInRequest.enableVaultManager = true
    dropInRequest.allowVaultCardOverride = true
    dropInRequest.vaultCardDefaultValue = true
    dropInRequest.cardholderNameStatus = FIELD_OPTIONAL
    dropInRequest.vaultVenmoDefaultValue = true
```

> The full list of changed parameters is below:

Author: Open to suggestions on any of the parameter name changes

1. `clientTokent` -> removed
1. `tokenizationKey` -> removed
1. `amount` -> removed
1. `intent` -> removed
1. `collectDeviceData` -> `shouldCollectDeviceData`
1. `googlePaymentRequest` -> `googlePayRequest`
1. `disableGooglePayment` -> `googlePayDisabled`
1. `paypalRequest` -> `payPalRequest`
1. `disablePayPal` -> `payPalDisabled`
1. `disableVenmo` -> `venmoDisabled`
1. `disableCard` -> `cardDisabled`

> 1. `requestThreeDSecureVerification` -> `shouldRequestThreeDSecureVerification`

Reviewer: I think `should` and `Request` is redundant

Reviewer 2: I do like the shorter name. Also indifferent here.

1. `maskCardNumber` -> `shouldMaskCardNumber`
1. `maskSecurityCode` -> `shouldMaskSecurityCode`
1. `vaultManager` -> `enableVaultManager`
1. `vaultCard` -> `vaultCardDefaultValue`
1. `vaultVenmo` -> `vaultVenmoDefaultValue`

> 

Reviewer: Hm - in general, I don't love `should`. To me, `should` makes sense it if it possible that the thing they're requesting for might not happen.

Reviewer: We do use should in a few places in iOS, currently. For example we use it for  `shouldMaskSecurityCode`. Up to you if you want to keep that in parity.

Reviewer 2: Yeah I think that naming pattern is common in iOS. I don't think there's another reason Java/Kotlin interop only sees `is` as a known convention for booleans when synthesizing property accessors.

## Authorization

The `clientToken`, `tokenizationKey`, and `authorization` fields have been removed from `DropInRequest`. 

> In v6, authorization should be included when instantiating a `DropInClient` instead.

Reviewer: Can we include a code snippet here?

Reviewer: Also, maybe this section should go below the section where we introduce `DropInClient`. Also, might be helpful if `DropInClient` has it's own section. We could have a code snippet comparing the old to new integration pattern so they can be clear how they should migrate away from `BraintreeFragment`.

Author: I added a small section for `DropInClient` without any code snippets since they would be the same snippets from the authorization and the launch drop in sections. I think it makes sense now that a reader learns about the client, sees how to instantiate with authorization, and then sees how to launch Drop-in. `BraintreeFragment` wasn't a merchant-facing part of Drop-in, so I don't think it makes sense to put that in the migration guide here, but I added a note about no longer needing an `Intent` to launch Drop-in. Let me know if this looks better.

Reviewer: Ah gotcha - thanks for clarifying!

## Launch Drop-In 

`DropInClient` has been added in v6 and is responsible for launching `DropInActivity`. 
To launch Drop-In, instantiate a `DropInClient` and call `DropInClient#launchDropInForResult`:

Java:
```java
DropInClient dropInClient = new DropInClient(this, "TOKENIZATION_KEY_OR_CLIENT_TOKEN", dropInRequest);
dropInClient.launchDropInForResult(this, DROP_IN_REQUEST_CODE);
```

Kotlin:
```kotlin
val dropInClient = DropInClient(this, "TOKENIZATION_KEY_OR_CLIENT_TOKEN", dropInRequest)
dropInClient.launchDropInForResult(this, DROP_IN_REQUEST_CODE)
```

## Handle Drop-In Result

> Handle the result from Drop-In in `onActivityResult` as in v5. 

Reviewer: ```suggestion
You should handle the `<TYPE>` in `onActivityResult`, the same was as you did in your v5 integration.
```
Take it or leave it

> The key for accessing an error from the Activity result has moved from `DropInActivity.EXTRA_ERROR` to `DropInResult.EXTRA_ERROR`.
> The method for accessing the `String` payment method nonce from `DropInResult` has been updated from `PaymentMethodNonce#getNonce` to `PaymentMethodNonce#getString`. 

Reviewer: These get lost a bit. Maybe we can do something like:

```suggestion

Changes:
- The key for accessing an error from the Activity result has changed from `DropInActivity.EXTRA_ERROR` to `DropInResult.EXTRA_ERROR`.
- The method for accessing the `String` payment method nonce from `DropInResult` has been updated from `PaymentMethodNonce#getNonce` to `PaymentMethodNonce#getString`. 
```
